File: python/백준/구현/14890 경사로.py
import sys

input = sys.stdin.readline

N, L = map(int, input().split())
board = []
for _ in range(N):
    board.append(list(map(int, input().split())))

def check(line):
    used = [False] * N
    for i in range(N - 1):
        diff = line[i+1] - line[i]
        if diff == 0:
            continue
        elif diff == 1:
            for j in range(i, i - L, -1):
                if j < 0 or line[j] != line[i] or used[j]:
                    return False
                used[j] = True
        elif diff == -1:
            for j in range(i + 1, i + 1 + L):
                if j >= N or line[j] != line[i+1] or used[j]:
                    return False
                used[j] = True
        else:
            return False
    return True

result = 0

for i in range(N):
    if check(board[i]):
        result += 1
    col = [board[r][i] for r in range(N)]
    if check(col):
        result += 1

print(result)

File: cpp/백준/구현/14890 경사로.cpp
#include <iostream>
#include <vector>
using namespace std;

int N, L;
vector<vector<int>> board;

bool check(const vector<int>& line) {
    vector<bool> used(N, false);
    for (int i = 0; i < N - 1; ++i) {
        int diff = line[i + 1] - line[i];
        if (diff == 0) continue;
        else if (diff == 1) {
            for (int j = i; j > i - L; --j) {
                if (j < 0 || line[j] != line[i] || used[j]) return false;
                used[j] = true;
            }
        } else if (diff == -1) {
            for (int j = i + 1; j < i + 1 + L; ++j) {
                if (j >= N || line[j] != line[i + 1] || used[j]) return false;
                used[j] = true;
            }
        } else {
            return false;
        }
    }
    return true;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> N >> L;

    board.assign(N, vector<int>(N));
    for (int i = 0; i < N; ++i)
        for (int j = 0; j < N; ++j)
            cin >> board[i][j];

    int result = 0;

    for (int i = 0; i < N; ++i) {
        if (check(board[i])) result++;
        vector<int> col(N);
        for (int r = 0; r < N; ++r) col[r] = board[r][i];
        if (check(col)) result++;
    }

    cout << result << "\n";
    return 0;
}
